// node-red-contrib-cache/dist/cache.js
'use strict';

function _interopRequireDefault(e) {
  return e && e.__esModule ? e : {
    'default': e
  };
}

function _classCallCheck(e, t) {
  if (!(e instanceof t)) throw new TypeError('Cannot call a class as a function');
}

Object.defineProperty(exports, '__esModule', {
  'value': !0
});

var _nodeCache = require('node-cache'),
    _nodeCache2 = _interopRequireDefault(_nodeCache);

exports.default = function (e) {
  var t = function t(c) {
    function a(e) {
      this.nodeList.push(e);
    }

    function s() {
      this.nodeList.forEach(function (e) {
        e.emit('updated');
      });
    }

    var o = this;
    _classCallCheck(this, t), e.nodes.createNode(this, c), this.name = c.name, this.cache = new _nodeCache2.default({
      'stdTTL': c.defaultTtl || 0,
      'checkperiod': c.checkPeriod || 0
    }), this.cache.nodeList = [], this.cache.addNode = a.bind(this.cache), this.cache.onChanged = s.bind(this.cache), ['set', 'del', 'expired', 'flush'].forEach(function (e) {
      o.cache.on(e, o.cache.onChanged);
    }), this.on('close', function () {
      o.cache.close(), delete o.cache;
    });
  };

  e.nodes.registerType('Cache', t);

  var c = function t(c) {
    var a = this;
    _classCallCheck(this, t), e.nodes.createNode(this, c), this.keyProperty = c.keyProperty || 'topic', this.valueProperty = c.valueProperty || 'payload', this.useString = c.useString, this.cacheMissRouting = c.outputs > 1, this.cacheNodeId = c.cache, this.cacheNode = e.nodes.getNode(this.cacheNodeId), this.cacheNode && (this.cacheNode.cache.addNode(this), this.on('updated', function () {
      a.status({
        'fill': 'green',
        'shape': 'dot',
        'text': e._('cache.status.keys', {
          'n': a.cacheNode.cache.getStats().keys
        })
      });
    })), this.name = c.name;

    var s = function (e, t) {
      if (a.cacheMissRouting) {
        var c = [];
        c[t ? 1 : 0] = e, a.send(c);
      } else a.send(e);
    };

    this.on('input', function (t) {
      if (a.cacheNode) if (t.dump || t.payload && t.payload.dump) a.cacheNode.cache.keys(function (c, o) {
        c || a.cacheNode.cache.mget(o, function (c, o) {
          e.util.setMessageProperty(t, a.valueProperty, o), s(t);
        });
      });else {
        var c = e.util.getMessageProperty(t, a.keyProperty);
        c && a.cacheNode.cache.get(c, function (c, o) {
          if (!c) {
            var i = void 0 === o;
            e.util.setMessageProperty(t, a.valueProperty, '' === o || i ? null : o), s(t, i);
          }
        });
      }
    }), process.nextTick(function () {
      a.emit('updated');
    });
  };

  e.nodes.registerType('Cache in', c);

  var a = function t(c) {
    var a = this;
    _classCallCheck(this, t), e.nodes.createNode(this, c), this.keyProperty = c.keyProperty || 'topic', this.valueProperty = c.valueProperty || 'payload', this.ttlProperty = c.ttlProperty || '', this.useString = c.useString, this.cacheNodeId = c.cache, this.cacheNode = e.nodes.getNode(this.cacheNodeId), this.cacheNode && (this.cacheNode.cache.addNode(this), this.on('updated', function () {
      a.status({
        'fill': 'green',
        'shape': 'dot',
        'text': e._('cache.status.keys', {
          'n': a.cacheNode.cache.getStats().keys
        })
      });
    })), this.name = c.name, this.on('input', function (t) {
      if (a.cacheNode) {
        var c = e.util.getMessageProperty(t, a.keyProperty);

        if (c) {
          var s = e.util.getMessageProperty(t, a.valueProperty);

          if (a.ttlProperty) {
            var o = e.util.getMessageProperty(t, a.ttlProperty) || 0;
            a.cacheNode.cache.set(c, s, o);
          } else a.cacheNode.cache.set(c, s);
        }
      }
    }), process.nextTick(function () {
      a.emit('updated');
    });
  };

  e.nodes.registerType('Cache out', a);
}, module.exports = exports.default;
